allennlp/data/dataset_readers/penn_tree_bank.py

```diff
@@ -0,0 +1,173 @@
+        pos_tag_field = SequenceLabelField(pos_tags, text_field, "pos_tags")
+        fields = {"tokens": text_field, "pos_tags": pos_tag_field}
+        spans: List[Field] = []
+        gold_labels = []
+
+        if gold_tree is not None:
+            gold_spans_with_pos_tags = self._get_gold_spans(gold_tree, 0, OrderedDict()).items()
+            gold_spans = {span: label for (span, label)
+                          in gold_spans_with_pos_tags if "-POS" not in label}
+        else:
+            gold_spans = None
+        for start, end in enumerate_spans(tokens):
+            spans.append(SpanField(start, end, text_field))
+
+            if gold_spans is not None:
+                if (start, end) in gold_spans.keys():
+                    gold_labels.append(gold_spans[(start, end)])
+                else:
+                    gold_labels.append("NO-LABEL")
+
+        span_list_field: ListField = ListField(spans)
+        fields["spans"] = span_list_field
+        if gold_tree is not None:
+            fields["span_labels"] = SequenceLabelField(gold_labels, span_list_field)
+
+        return Instance(fields)
+
+    def _get_gold_spans(self, # pylint: disable=arguments-differ
+                        tree: Tree,
+                        index: int,
+                        typed_spans: OrderedDict) -> OrderedDict:
+        """
+        Recursively construct the gold spans from an nltk ``Tree``.
+        Spans are inclusive.
+
+        Parameters
+        ----------
+        tree : ``Tree``, required.
+            An NLTK parse tree to extract spans from.
+        index : ``int``, required.
+            The index of the current span in the sentence being considered.
+        typed_spans : ``OrderedDict[Tuple[int, int], str]``, required.
+            A ordered dictionary mapping spans to span labels.
+
+        Returns
+        -------
+        typed_spans : ``OrderedDict[Tuple[int, int], str]``, required.
+            A ordered dictionary mapping all subtree spans in the parse tree
+            to their constituency labels. Leaf nodes have POS tag spans, which
+            are denoted by a label of "LABEL-POS".
+        """
+        # NLTK leaves are strings.
+        if isinstance(tree[0], str):
+            # The "length" of a tree is defined by
+            # NLTK as the number of children.
+            # We don't actually want the spans for leaves, because
+            # their labels are pos tags. However, it makes the
+            # indexing more straightforward, so we'll collect them
+            # and filter them out below. We subtract 1 from the end
+            # index so the spans are inclusive.
+            typed_spans[(index, index + len(tree) - 1)] = tree.label() + "-POS"
+        else:
+            # otherwise, the tree has children.
+            first = True
+            for child in tree:
+                typed_spans = self._get_gold_spans(child, index, typed_spans)
```

Comment:
Overriding this variable, instead of assigning the return value to a new variable, is confusing to me, and seems error-prone.

On the other hand, it looks like you require this behavior, because of the way the recursive function collects its result.  I'd recommend just omitting the return value here, and doing something like:

```python
for child in tree:
    # We're ignoring the return value here because this call modifies typed_spans directly.
    self._get_gold_spans(child, index, typed_spans)
    ...
```